gnovm/pkg/gnolang/garbage_collector.go

```diff
@@ -42,65 +74,108 @@ func (m *Machine) GarbageCollect() (left int64, ok bool) {
 	}
 
 	// Visit exceptions
-	for exception := range m.Exceptions {
-		// XXX implement for exceptions.
-		// XXX Exception is not an object,
-		// so implement a custom method and pass in vis.
+	for _, exception := range m.Exceptions {
+		stop = exception.Visit(vis, m.Store)
+		if stop {
+			return -1, false
+		}
 	}
 
 	// Return bytes remaining.
 	maxBytes, bytes := m.Alloc.Status()
 	return maxBytes - bytes, true
 }
 
-// Returns a visitor that bumps the gcCycle counter
+// Returns a visitor that bumps the GCCycle counter
 // and stops if alloc is out of memory.
-func GCVisitorFn(gcCycle int64, alloc *Alloc) Visitor {
-	vis := func(o Object) bool {
-		// Return if already measured.
-		if o.GetLastGCCycle() == gcCycle {
-			return false // but don't stop
+func GCVisitorFn(gcCycle int64, alloc *Allocator) Visitor {
+	var vis func(value Value) bool
+
+	vis = func(v Value) bool {
+		if v == nil || v == (*Block)(nil) || v == (*PackageValue)(nil) {
+			return false
+		}
+
+		if debug {
+			debug.Printf("Visit, v: %v (type: %v)\n", v, reflect.TypeOf(v))
+		}
+
+		oo, isObject := v.(Object)
+
+		if isObject {
+			// Return if already measured.
+			if debug {
+				debug.Printf("oo.GetLastGCCycle: %d, gcCycle: %d\n", oo.GetLastGCCycle(), gcCycle)
+			}
+			if oo.GetLastGCCycle() == gcCycle {
+				return false // but don't stop
+			}
 		}
+
+		alloc.visitCount++ // Count operations for gas calculation
+
 		// Add object size to alloc.
-		size := o.GetShallowSize()
+		size := v.GetShallowSize()
 		alloc.Allocate(size)
+
 		// Stop if alloc max exceeded.
+		// NOTE: Unlikely to occur, but keep it here for
+		// now to handle potential edge cases.
+		// Consider removing it later if no issues arise.
 		maxBytes, curBytes := alloc.Status()
 		if maxBytes < curBytes {
 			return true
 		}
-		// Invote the traverser on o.
-		stop := o.VisitAssociated(vis)
+
+		// Invoke the traverser on v.
+		stop := v.VisitAssociated(vis)
+
 		// Finally bump cycle.
-		o.SetLastGCCycle(gcCycle)
+		if isObject {
+			oo.SetLastGCCycle(gcCycle)
+		}
```

Comment:
maybe a defer on top so we don't forget it

Reply:
done.